### PR TITLE
cagebreak: 1.6.0 -> 1.7.0

### PR DESCRIPTION
**Pull request by Synthetica9**
###### Motivation for this change
Tests still failing, related to the failure in https://github.com/NixOS/nixpkgs/pull/119615

###### Things done

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

**Comment by Synthetica9** [comment]
But in this one I got an actual backtrace with `-fsanitize=address`!

```
00:00:00.572 [ERROR] [backend/drm/atomic.c:35] connector DP-1: Atomic commit failed (modeset): Invalid argument
00:00:01.848 [ERROR] [../cagebreak.c:186] Could not open config file "/home/synthetica/.config/cagebreak/config"
=================================================================
==88383==ERROR: AddressSanitizer: heap-use-after-free on address 0x61500002df88 at pc 0x000000420744 bp 0x7ffed0623660 sp 0x7ffed0623658
READ of size 8 at 0x61500002df88 thread T0
    #0 0x420743 in seat_set_focus (/nix/store/r1n5267ndk9rf4qcp55lkjkid1mkm0ij-cagebreak-1.7.0/bin/.cagebreak-wrapped+0x420743)
    #1 0x40eadf in keybinding_cycle_outputs (/nix/store/r1n5267ndk9rf4qcp55lkjkid1mkm0ij-cagebreak-1.7.0/bin/.cagebreak-wrapped+0x40eadf)
    #2 0x41518f in output_clear (/nix/store/r1n5267ndk9rf4qcp55lkjkid1mkm0ij-cagebreak-1.7.0/bin/.cagebreak-wrapped+0x41518f)
    #3 0x415343 in output_destroy (/nix/store/r1n5267ndk9rf4qcp55lkjkid1mkm0ij-cagebreak-1.7.0/bin/.cagebreak-wrapped+0x415343)
    #4 0x7f093ebc144b in wlr_signal_emit_safe (/nix/store/jvkgicprzd0sq9akf6kjx9fgcyfljibz-wlroots-0.13.0/lib/libwlroots.so.8+0x7744b)
    #5 0x7f093ebad7e8 in wlr_output_damage_destroy (/nix/store/jvkgicprzd0sq9akf6kjx9fgcyfljibz-wlroots-0.13.0/lib/libwlroots.so.8+0x637e8)
    #6 0x7f093ebc144b in wlr_signal_emit_safe (/nix/store/jvkgicprzd0sq9akf6kjx9fgcyfljibz-wlroots-0.13.0/lib/libwlroots.so.8+0x7744b)
    #7 0x7f093ebb31b9 in wlr_output_destroy (/nix/store/jvkgicprzd0sq9akf6kjx9fgcyfljibz-wlroots-0.13.0/lib/libwlroots.so.8+0x691b9)
    #8 0x7f093eb7f934 in destroy_drm_connector (/nix/store/jvkgicprzd0sq9akf6kjx9fgcyfljibz-wlroots-0.13.0/lib/libwlroots.so.8+0x35934)
    #9 0x7f093eb7c49c in backend_destroy.part.0 (/nix/store/jvkgicprzd0sq9akf6kjx9fgcyfljibz-wlroots-0.13.0/lib/libwlroots.so.8+0x3249c)
    #10 0x7f093eb87964 in multi_backend_destroy (/nix/store/jvkgicprzd0sq9akf6kjx9fgcyfljibz-wlroots-0.13.0/lib/libwlroots.so.8+0x3d964)
    #11 0x7f093ec10dbc in wl_display_destroy (/nix/store/7fnq0qvpvgj04gdf2l6hmgpz8dk72r1z-wayland-1.19.0/lib/libwayland-server.so.0+0x8dbc)
    #12 0x408da4 in main (/nix/store/r1n5267ndk9rf4qcp55lkjkid1mkm0ij-cagebreak-1.7.0/bin/.cagebreak-wrapped+0x408da4)
    #13 0x7f093e96cdec in __libc_start_main (/nix/store/1jn6apz0fa9h9x7rl3v6vwiymwnjznwv-glibc-2.32-40/lib/libc.so.6+0x27dec)
    #14 0x40a429 in _start (/nix/store/r1n5267ndk9rf4qcp55lkjkid1mkm0ij-cagebreak-1.7.0/bin/.cagebreak-wrapped+0x40a429)

0x61500002df88 is located 8 bytes inside of 504-byte region [0x61500002df80,0x61500002e178)
freed by thread T0 here:
    #0 0x7f093f112b6f in __interceptor_free (/nix/store/hsw0sq8y0a46fxc2d7krgyr6p5dy71kd-gcc-10.2.0-lib/lib/libasan.so.6+0xacb6f)
    #1 0x7f093ebc144b in wlr_signal_emit_safe (/nix/store/jvkgicprzd0sq9akf6kjx9fgcyfljibz-wlroots-0.13.0/lib/libwlroots.so.8+0x7744b)
    #2 0x60f0000088bf  (<unknown module>)

previously allocated by thread T0 here:
    #0 0x7f093f113037 in __interceptor_calloc (/nix/store/hsw0sq8y0a46fxc2d7krgyr6p5dy71kd-gcc-10.2.0-lib/lib/libasan.so.6+0xad037)
    #1 0x41f7d2 in seat_create (/nix/store/r1n5267ndk9rf4qcp55lkjkid1mkm0ij-cagebreak-1.7.0/bin/.cagebreak-wrapped+0x41f7d2)
    #2 0x60d00000044f  (<unknown module>)

SUMMARY: AddressSanitizer: heap-use-after-free (/nix/store/r1n5267ndk9rf4qcp55lkjkid1mkm0ij-cagebreak-1.7.0/bin/.cagebreak-wrapped+0x420743) in seat_set_focus
Shadow bytes around the buggy address:
  0x0c2a7fffdba0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c2a7fffdbb0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c2a7fffdbc0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c2a7fffdbd0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 fa fa
  0x0c2a7fffdbe0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
=>0x0c2a7fffdbf0: fd[fd]fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2a7fffdc00: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2a7fffdc10: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
  0x0c2a7fffdc20: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fa
  0x0c2a7fffdc30: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c2a7fffdc40: fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd fd
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==88383==ABORTING
```

**Comment by legendofmiracles** [comment]
Result of `nixpkgs-review pr 119963` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built:</summary>
  <ul>
    <li>cagebreak</li>
  </ul>
</details>

**Comment by r-rmcgibbo** [comment]
Result of `nixpkgs-review pr 119963` at b6ba402a run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
  <summary>1 package built successfully:</summary>
  <ul>
    <li>cagebreak</li>
  </ul>
</details>
<details>
  <summary>1 suggestion:</summary>
  <ul>
    <li>warning: <a href="https://github.com/jtojnar/nixpkgs-hammering/blob/master/explanations/build-tools-in-build-inputs.md">build-tools-in-build-inputs</a>

pandoc is a build tool so it likely goes to `nativeBuildInputs`, not `buildInputs`.

Near pkgs/applications/window-managers/cagebreak/default.nix:26:3:
```
   |
26 |   buildInputs = [
   |   ^
```
</li>
  </ul>
</details>

File: pkgs/applications/window-managers/cagebreak/default.nix [hunk 1]
```diff
@@ -3,25 +3,29 @@
...
   pname = "cagebreak";
-  version = "1.6.0";
+  version = "1.7.0";
 
   src = fetchFromGitHub {
     owner = "project-repo";
     repo = "cagebreak";
     rev = version;
-    hash = "sha256-F7fqDVbJS6pVgmj6C1/l9PAaz5yzcYpaq6oc6a6v/Qk=";
+    sha256 = "10ps59dbvxalkdy0rs67k4dgkhy16xspdwpxl551dcxrgcj2740y";
```
Comment (berbiche) [inline pkgs/applications/window-managers/cagebreak/default.nix:17]: I thought there was a general move towards replacing `sha256` with the new hashes from Nix 2.4?

Comment (Synthetica9) [reply pkgs/applications/window-managers/cagebreak/default.nix:17]: I didn't get the memo on that one, source?

Comment (berbiche) [reply pkgs/applications/window-managers/cagebreak/default.nix:17]: I don't have any link. I came to this understanding from discussions in \#nixos on IRC.

Anyhow, the new hashes are only generated when using nix 2.4 tooling IIRC (`nixUnstable`) so it doesn't make sense to move to something that has not been stabilized yet.

**Comment by berbiche** [comment]
@Synthetica9 IIRC cagebreak will crash on startup if the config file is missing (poor UX choice imo)

**Comment by Synthetica9** [comment]
> @Synthetica9 IIRC cagebreak will crash on startup if the config file is missing (poor UX choice imo)

Oh, yeah, that's very poor UX, but thanks for the heads-up

File: pkgs/applications/window-managers/cagebreak/default.nix (hunk 1 above)
Comment (SuperSandro2000) [reply pkgs/applications/window-managers/cagebreak/default.nix:17]: Right now it does not matter which hash you are using. Just use the one that your nix-daemon provides.

> Anyhow, the new hashes are only generated when using nix 2.4 tooling IIRC (nixUnstable) so it doesn't make sense to move to something that has not been stabilized yet.


If you are using a newer nix-daemon then you would need to convert it every time which is tedious and not required.

File: nixos/tests/cagebreak.nix [hunk 2]
```diff
@@ -56,6 +56,10 @@ in
       '';
     };
 
+    # Normal QEMU drivers don't work after wlroots 0.13 it seems, fix ported
+    # from #119615
+    environment.variables.QEMU_OPTS = "-vga virtio";
```
[the remark this answers is not in the document]
Comment (primeos) [reply nixos/tests/cagebreak.nix:61]: Oh, now I see :)
I don't think this should work as `QEMU_OPTS` is required on the host for the test driver script that launches the VM (see `nixos/lib/test-driver/test-driver.py`).

[the remark this answers is not in the document]
Comment (Synthetica9) [reply nixos/tests/cagebreak.nix:61]: I agree that it's rather mystifying, but it _does_ work (test passes with this line and fails without this line)

[the remark this answers is not in the document]
Comment (Synthetica9) [reply nixos/tests/cagebreak.nix:61]: Wait, now it doesn't pass again. I have no idea what's going on here, but I guess I'll use the other method

File: nixos/tests/cagebreak.nix [hunk 3]
```diff
@@ -56,6 +56,7 @@ in
       '';
     };
 
+
```
Comment (primeos) [inline nixos/tests/cagebreak.nix:59]: Seems like this line was introduced on accident? (not a blocker but might make sense to remove)

**Review by primeos — CHANGES_REQUESTED** [review]
You might be already on it but could you also rebase to resolve the merge conflict and limit the test to `x86_64-linux` (https://github.com/NixOS/nixpkgs/commit/28b8cff301f356a7a24be1b6203f6a2e3a814e7d#diff-307c155c790abf7b54b5dec19dc230c6d692c6dca0604760d8389c7aed543c10R50)?

**Comment by primeos** [comment]
Yesterday I had a brief look at this and got the following output for `nix-build -A nixosTests.cagebreak`:
```
machine # [    9.081583] xsession[933]: 00:00:00.002 [ERROR] [backend/x11/backend.c:433] X11 does not support DRI3 extension
machine # [    9.083267] xsession[933]: 00:00:00.004 [ERROR] [../cagebreak.c:313] Unable to create the wlroots backend
machine # [    9.196526] show_signal_msg: 77 callbacks suppressed
machine # [    9.196542] .cagebreak-wrap[933]: segfault at 8 ip 0000000000408574 sp 00007ffd441af370 error 4 in .cagebreak-wrapped[407000+d000]
machine # [    9.198950] Code: f4 ff ff 4c 8b 84 24 70 01 00 00 8d 45 01 48 89 c5 49 8b 3c c0 48 85 ff 75 e4 4c 89 c7 e8 84 f4 ff ff 48 8b bc 24 18 01 00 00 <48> 8b 47 08 4c 8d 6f d8 48 8d 68 d8 48 39 df 75 0e eb 36 66 0f 1f
machine # [    9.107703] systemd[1]: Created slice system-systemd\x2dcoredump.slice.
machine # [    9.111248] systemd[1]: Started Process Core Dump (PID 935/UID 0).
machine # [    9.192686] systemd-coredump[936]: Process 933 (.cagebreak-wrap) of user 1000 dumped core.
machine # [    9.202025] systemd[1]: systemd-coredump@0-935-0.service: Succeeded.
machine # [    9.213734] xsession[925]: /nix/store/di4ylzjkpqg9v3jzp6d2y43lscm3wj36-xsession: line 21:   933 Segmentation fault      (core dumped) /nix/store/sfhniimqs51gp2blcrqsya6rli08dpb4-cagebreak-1.7.0/bin/cagebreak
machine # [    9.227817] systemd[918]: Stopped target Current graphical user session.
machine # [    9.233687] lightdm[913]: pam_unix(lightdm-autologin:session): session closed for user alice
machine # [    9.239386] systemd-logind[559]: Session 1 logged out. Waiting for processes to exit.
machine # [    9.246370] systemd[1]: session-1.scope: Succeeded.
machine # [    9.247607] systemd-logind[559]: Removed session 1.
machine # [    9.800322] systemd-logind[559]: New session c1 of user lightdm.
```

The first message (`X11 does not support DRI3 extension`) comes from wlroots and got my attention. It seems like wlroots 0.13 started requiring DRI3 for the X11 backend and apparently that isn't supported for virtual GPU drivers (which kinda makes sense although `virtio` is pretty fast and supports 3D acceleration). I haven't analyzed this further though.

Starting Cagebreak directly is likely the best solution but I haven't looked much into Wayland tests so far (wanted to investigate manually logging in via a VT/tty for a Sway test but never got around to that).

Note: I'll rebase this PR to resolve the merge conflict (2da39dda3ead24769696e2a7589d40b4f43a4ff2).

**Pull request closed without merging**